gmail/src/models/state.ts

```diff
@@ -0,0 +1,196 @@
+        const sharedSecret = scriptProperties.getProperty("ODOO_SHARED_SECRET");
+        return this._isTrue(sharedSecret);
+    }
+
+    /**
+     * Dictionary which inform us on which record we already logged the email.
+     * So the user can not log 2 times the same email on the same record.
+     * This is stored into the cache, so we don't need to modify the Odoo models.
+     *
+     * Note: the cache expire after 6 hours.
+     *
+     * Returns:
+     *     {
+     *         "partners": [3, 6], // email already logged on the partner 3 and 6
+     *         "leads": [7, 14],
+     *     }
+     */
+    static getLoggingState(messageId: string) {
+        const cache = CacheService.getUserCache();
+        const loggingStateStr = cache.get("ODOO_LOGGING_STATE_" + this.odooServerUrl + "_" + messageId);
+
+        const defaultValues: Record<string, number[]> = {
+            partners: [],
+            leads: [],
+            tickets: [],
+        };
+
+        if (!loggingStateStr || !loggingStateStr.length) {
+            return defaultValues;
+        }
+        return { ...defaultValues, ...JSON.parse(loggingStateStr) };
+    }
+
+    /**
+     * Save the fact that we logged the email on the record, in the cache.
+     *
+     * Returns:
+     *     True if the record was not yet marked as "logged"
+     *     False if we already logged the email on the record
+     */
+    static setLogged(messageId: string, res_model: string, res_id: number): boolean {
+        const loggingState = this.getLoggingState(messageId);
+        if (loggingState[res_model].indexOf(res_id) < 0) {
+            loggingState[res_model].push(res_id);
+            const cache = CacheService.getUserCache();
+
+            // The cache key depend on the current email open and on the Odoo database
+            const cacheKey = "ODOO_LOGGING_STATE_" + this.odooServerUrl + "_" + messageId;
+
+            cache.put(
+                cacheKey,
+                JSON.stringify(loggingState),
+                21600, // 6 hours, maximum cache life time
+            );
+            return true;
+        }
+        return false;
+    }
+
+    /* DEBUG */
+    static get debug() {
+        const userProperties = PropertiesService.getUserProperties();
+        const debugStr = userProperties.getProperty("ODOO_DEBUG");
+        if (this._isTrue(debugStr)) {
+            const debug = JSON.parse(debugStr);
+            debug.httpCallCount = debug.httpCallCount || 0;
+            return debug;
+        }
+        return {};
+    }
+
+    static set debug(value: any) {
+        const userProperties = PropertiesService.getUserProperties();
+        const debugStr = userProperties.setProperty("ODOO_DEBUG", JSON.stringify(value));
+    }
+
+    static debugIncrementHttpCallCount() {
+        const debug_ = this.debug;
+
+        if (!debug_.httpCallCount) {
+            debug_.httpCallCount = 0;
+        }
+        debug_.httpCallCount += 1;
+
+        this.debug = debug_;
+    }
```

Comment:
What is all this?

Reply:
It's only use for debug purpose, to know how many HTTP request we made (and so if we read in the cache or not when contacting IAP), but I can remove it I guess :)